### ISSUE TITLE
[feature request] RetroAchievements compatibility

### ISSUE THREAD

**Issue by meleu** [libretro/libretro-fceumm#52]
Retroarch recently got some really nice fixes related to the RetroAchievements feature and they solved most of the problems. But a few NES cheevos aren't working with FCEUmm, while working just fine with QuickNES. Our fellow @leiradel said that it happens because FCEUmm expose the Save RAM in a way that doesn't work with some achievements.

It would be really cool if FCEUmm became fully compatible with cheevos (I like the turbo buttons and [QuickNES is currently buggy on Android](https://github.com/libretro/RetroArch/issues/3973)).

If it helps, here are some save states for Metroid that aren't working with FCEUmm but works just fine with QuickNES: 
[Metroid (USA).savestates.zip](https://github.com/libretro/libretro-fceumm/files/589904/Metroid.USA.savestates.zip) *(the saves were taken using FCEUmm)*

The problems are with the pickup items only, all the other Metroid cheevos are working fine.

I noticed that the [Goomba Massacre](http://retroachievements.org/Achievement/4622) cheevo, from Super Mario Bros 3 also doesn't work with FCEUmm and works fine with QuickNES.

**Comment by leiradel** [comment]
Neither of these two cores implement the memory map interface, so cheevos are supported via the `retro_get_memory` functions.

QuickNES returns valid data for the Save RAM, but FCEUmm doesn't:

```
QuickNES

RetroArch [INFO] :: CHEEVOS system RAM: 00000000052A4460 2048
RetroArch [INFO] :: CHEEVOS save RAM:   00000000052BEF60 8192

FCEUmm

RetroArch [INFO] :: CHEEVOS system RAM: 0000000070CC07C0 2048
RetroArch [INFO] :: CHEEVOS save RAM:   0000000000000000 0
```

Cheevos that require the Save RAM won't work with FCEUmm. It's odd because there is code in it to return something for the Save RAM.

Both return `NULL` and 0 for the Video RAM and RTC, but those aren't used in any cheevos so far.

**Comment by leiradel** [comment]
I believe it amounts to whether `iNESCart.battery` is `true` or `false`. Loading Galaga with FCEUmm prints this message:

```
RetroArch [libretro INFO] ::  Battery-backed: No
```

So `SaveGame` and `SaveGameLen` are (possibly) not initialized, which makes `retro_get_memory` return `NULL` and 0.

**Comment by retro-wertz** [comment]
with regards to this, i am wondering if RETRO_ENVIRONMENT_SET_SUPPORT_ACHIEVEMENTS can be set on this and announce this core to be retroarchievement-compatible.

**Comment by leiradel** [comment]
Strange, it seems someone have added complete cheevos support for FCEUmm more than an year ago: https://github.com/libretro/libretro-fceumm/commit/613698e7944baf707f69981b398f7f698b5cb94d

The log shows that the Save RAM is indeed there:

```
[INFO] CHEEVOS system RAM: 0000000070CC8AE0 2048
[INFO] CHEEVOS save RAM:   0000000006BA4610 8192
```

Can someone confirm if the problematic cheevos are working? Thanks!

**Comment by retro-wertz** [comment]
which problematic ones? i did prefer fceumm for cheevos and it did update
achievements but not particular of what issue were there if any.

**Comment by leiradel** [comment]
The OP mentioned two cheevos that weren't being awarded with FCEUmm.

**Comment by retro-wertz** [comment]
can only test with metroid, and its not working in fceumm as stated.

**Comment by retro-wertz** [comment]
UPDATE: just a *fun* test here.
assigning WRAM and its size to something like 
```
info->workramptr = WRAM;
info->workramsize = 8192;
```
and then calling this in retro_get_memory_data/retro_get_memory_size does the same fix.

**Comment by leiradel** [comment]
Thanks @retro-wertz, I think now I know what the problem is, at least a part of it.

So the NES has this ExWRAM, and there are even comments about it in FCEUmm's libretro port source code:

```C++
// TODO: add ExWRAM(0x2000) support ?
```

The regular Work RAM sits in the space from 0x0000 to 0x07ff, and is mirrored to fill the remaining space from 0x0800 to 0x1ffff. Looking at the NES memory layout, the only place where this extended Work RAM fits is in the space from 0x0000 to 0x1fff.

In RetroArch I had problems with cheevos accessing addresses in the mirrored RAM, so when the console is a NES I patch all the addresses with `addr &= 0x07fff`. This will obviously break legit reads from the ExRAM.

I could look to the Work RAM size, and only patch the addresses if its size of 0x0800. If it's 0x2000, it's the ExRAM, and I leave the addresses alone. However, Both emulators seem to always return 0x2000 for the Work RAM size, so maybe some patching for them will also be required.

**Comment by retro-wertz** [comment]
i think reading from WRAM is an alternative, if the rom has no battery-bit enabled. some mappers do have different WRAM size (8 default, there are 16 and 32)

**Comment by leiradel** [comment]
Hm, I think I don't have enough knowledge about the NES to properly solve this...

**Comment by meleu** [comment]
Hey guys, what literature do you recommend for a newbie willing to understand what you're saying?

I'm speaking serious. :) I wanna work out my low level programming skills. And I think that understanding the NES architecture is a good motivation. ;)

**Comment by meleu** [comment]
Wow! I think I've found something:
http://wiki.nesdev.com/w/index.php/NES_reference_guide

Sorry for the off-topic. :)

**Comment by leiradel** [comment]
I think I managed to make it work, with help from zeromus at #retroarch 

I tested all the Metroid savestates in the OP. If you guys can test it further with other problematic cheevos I appreciate. I'll wait for your feedback to make the PR for upstream.

https://github.com/leiradel/libretro-fceumm

**Comment by retro-wertz** [comment]
the other problematic cheevos is hard to achieve and needs practice(Mario
3), so im wondering if this is gonna be a replacement or just be used on
cases with roms without battery bit. i can test other stuff if this is
gonna be a replacement.

**Comment by leiradel** [comment]
I'm not sure about this battery thing. I expect FCEUmm to work with all cheevos now.

**Comment by meleu** [comment]
I'll try the Goomba Massacre with your fix.

**Comment by retro-wertz** [comment]
i was refering to this
https://github.com/leiradel/libretro-fceumm/blob/master/src/drivers/libretro/libretro.c#L1828

**Comment by leiradel** [comment]
With the use of the memory map interface, `retro_get_memory_data/size` aren't used anymore.

Look at `memset(descs, 0, sizeof(descs));` and the subsequent lines.

**Comment by retro-wertz** [comment]
ok noted. testing

**Comment by meleu** [comment]
Ha! Confirmed!! @leiradel Goomba Massacre works with your fceumm branch!

**Comment by retro-wertz** [comment]
nice

**Comment by leiradel** [comment]
Awesome, is it enough to submit the PR?

**Comment by meleu** [comment]
I think yes

**Comment by retro-wertz** [comment]
yeah looks ok now, im testing general cheevos though with
retro_get_memory_data/size disabled and cheevos still works.

**Comment by meleu** [comment]
Solved in #118 .
Thanks @leiradel ! ;-)
